libs/rs-sdk/seda-sdk-macros/src/lib.rs

```diff
@@ -0,0 +1,145 @@
+            item.sig.span(),
+            "Function must not be const.",
+        ));
+    }
+
+    if item.sig.asyncness.is_some() {
+        errors.push(syn::Error::new(
+            item.sig.span(),
+            "Function must not be async.",
+        ));
+    }
+
+    if !item.sig.generics.params.is_empty() {
+        errors.push(syn::Error::new(
+            item.sig.generics.span(),
+            "Function must not have generics or lifetimes.",
+        ));
+    }
+
+    if !item.sig.inputs.is_empty() {
+        errors.push(syn::Error::new(
+            item.sig.inputs.span(),
+            "Function must not take any arguments.",
+        ));
+    }
+
+    if item.sig.variadic.is_some() {
+        errors.push(syn::Error::new(
+            item.sig.variadic.span(),
+            "Function must not be variadic.",
+        ));
+    }
+
+    match &item.sig.output {
+        syn::ReturnType::Default => {}
+        _ => {
+            errors.push(syn::Error::new(
+                item.sig.output.span(),
+                "Function must not have a return type.",
+            ));
+        }
+    }
+}
+
+#[proc_macro_attribute]
+pub fn oracle_program(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(item as ItemImpl);
+
+    // Ensure only the execute and tally methods are present
+    let mut execute_fn = None;
+    let mut tally_fn = None;
+
+    let mut errors = Vec::new();
+    for item in &input.items {
+        match item {
+            syn::ImplItem::Const(item) => errors.push(syn::Error::new(
+                item.span(),
+                "Consts are not allowed to be defined in this macro.",
+            )),
+            syn::ImplItem::Fn(item) => {
+                if item.sig.ident == "execute" {
+                    execute_fn = Some(item);
+                } else if item.sig.ident == "tally" {
+                    tally_fn = Some(item);
+                } else {
+                    errors.push(syn::Error::new(
+                        item.span(),
+                        "Only the functions tally and execute are allowed.",
+                    ));
+                }
+            }
+            syn::ImplItem::Type(item) => errors.push(syn::Error::new(
+                item.span(),
+                "Only the type Error is allowed to be defined in this macro.",
+            )),
+            syn::ImplItem::Macro(item) => errors.push(syn::Error::new(
+                item.span(),
+                "Macros are not allowed to be defined in this macro.",
+            )),
+            syn::ImplItem::Verbatim(item) => errors.push(syn::Error::new(
+                item.span(),
+                "Verbatims are not allowed to be defined in this macro.",
+            )),
+            _ => unreachable!(),
+        }
+    }
+
+    if execute_fn.is_none() {
+        panic!("An 'execute' method is required.");
+    }
+
+    if tally_fn.is_none() {
+        panic!("A 'tally' method is required.");
+    }
```

Comment:
Can we allow having just one of the two?

Reply:
hmm a bit tricky but I think I can get that to work

Reply:
I managed. I put a `println` in the undefined function, since we still need the function with the way main is written. Is this fine or should I use the `Process::exit` or something else?

Reply:
I think process exit makes sense, we did that in the AS SDK as well to make it a little easier to notice that it's not implemented: `Process.error(Bytes.fromUtf8String("Tally not implemented"));`